crates/re_log_encoding/src/decoder/stream.rs

```diff
@@ -0,0 +1,281 @@
+
+    /// The stream state
+    state: State,
+}
+
+#[derive(Clone, Copy)]
+enum State {
+    StreamHeader,
+    MessageHeader,
+    Message(MessageHeader),
+}
+
+impl StreamDecoder {
+    pub fn new() -> Self {
+        Self {
+            compression: Compression::Off,
+            chunks: ChunkBuffer::new(),
+            uncompressed: Vec::with_capacity(1024),
+            state: State::StreamHeader,
+        }
+    }
+
+    pub fn push_chunk(&mut self, chunk: Vec<u8>) {
+        self.chunks.push(chunk);
+    }
+
+    pub fn try_read(&mut self) -> Result<Option<LogMsg>, DecodeError> {
+        match self.state {
+            State::StreamHeader => {
+                if let Some(header) = self.chunks.try_read(FileHeader::SIZE)? {
+                    // header contains version and compression options
+                    self.compression = read_options(header)?.compression;
+
+                    // we might have data left in the current chunk,
+                    // immediately try to read length of the next message
+                    self.state = State::MessageHeader;
+                    return self.try_read();
+                }
+            }
+            State::MessageHeader => {
+                if let Some(mut len) = self.chunks.try_read(MessageHeader::SIZE)? {
+                    let header = MessageHeader::decode(&mut len)?;
+                    self.state = State::Message(header);
+                    // we might have data left in the current chunk,
+                    // immediately try to read the message content
+                    return self.try_read();
+                }
+            }
+            State::Message(header) => {
+                if let Some(bytes) = self.chunks.try_read(header.compressed as usize)? {
+                    let bytes = match self.compression {
+                        Compression::Off => bytes,
+                        Compression::LZ4 => {
+                            self.uncompressed.resize(header.uncompressed as usize, 0);
+                            lz4_flex::block::decompress_into(bytes, &mut self.uncompressed)
+                                .map_err(DecodeError::Lz4)?;
+                            &self.uncompressed
+                        }
+                    };
+
+                    // read the message from the uncompressed bytes
+                    let message = rmp_serde::from_slice(bytes).map_err(DecodeError::MsgPack)?;
+
+                    self.state = State::MessageHeader;
+                    return Ok(Some(message));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+impl Default for StreamDecoder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+type Chunk = Cursor<Vec<u8>>;
+
+struct ChunkBuffer {
+    /// Any incoming chunks are queued until they are emptied
+    queue: VecDeque<Chunk>,
+
+    /// When `try_read` is called and we don't have enough bytes yet,
+    /// we store whatever we do have in this buffer
```

Comment:
The data is copied into the buffer, and the chunk retains it, but it won't be read ever again, because the `Chunk` cursor position is updated. The buffer exists so that `try_read` can return a contiguous slice.

Reply:
Add that as a docstring - future readers of the code will ask the same question!